### PR TITLE
TARIC Conformance Rules

### PR DESCRIPTION
**Pull request by matthewford**

### CODE REVIEW

File: app/validators/base_regulation_validator.rb [hunk 1]
```diff
@@ -20,26 +20,144 @@ class BaseRegulationValidator < TradeTariffBackend::Validator
...
+        (record.validity_end_date >= measure.validity_end_date))
+    end
+  end
+
   # TODO: ROIMB9
   # TODO: ROIMB10
-  # TODO: ROIMB11
+
+  validation :ROIMB11,
+             'The start date of the first PTS in time must be within the validity period of the base regulation. Apply the rule when changing the start date further forward in time.' do |record|
```
Comment (matthewford) [inline app/validators/base_regulation_validator.rb]: The rule says 'the first in time' which makes me think we need to update

https://github.com/bitzesty/trade-tariff-management/blob/02c91515373b39e534d802df8c1cd28dedfbc550/app/models/measure.rb#L104

to respect the time machine, not just pick the first

File: app/validators/base_regulation_validator.rb [hunk 2]
```diff
@@ -20,26 +20,144 @@ class BaseRegulationValidator < TradeTariffBackend::Validator
...
+    record.generating_measures.all? do |measure|
```
Comment (matthewford) [inline app/validators/base_regulation_validator.rb]: Also should we include the measure_partial_temporary_stops in the join for performance?

File: app/models/modification_regulation.rb [hunk 3]
```diff
@@ -37,4 +42,8 @@ def formatted_id
 
     "#{year}/#{number}"
   end
+
+  def last_fts_regulation
+    fts_regulations.last
```
Comment (matthewford) [inline app/models/modification_regulation.rb]: needs to use the time machine

**Comment by matthewford** [comment]
@macool needs rebasing

File: app/validators/base_regulation_validator.rb [hunk 4]
```diff
@@ -20,26 +20,144 @@ class BaseRegulationValidator < TradeTariffBackend::Validator
                         allow_blank: true
   end
 
-  # TODO: ROIMB5
-  # TODO: ROIMB6
+  validation :ROIMB5,
+             'If the regulation is replaced, completely or partially, modification is allowed only on the fields "Publication Date", "Official journal Number", "Official journal Page" and "Regulation Group Id".',
+             on: [:update] do |record|
```
[the remark this answers is not in the document]
Comment (mechanicles) [reply app/validators/base_regulation_validator.rb]: Yes, we cover this `record.modification_regulations.any?` condition for this validation. I think this regulation is fine.

File: app/models/base_regulation.rb [hunk 5]
```diff
@@ -7,13 +7,30 @@ class BaseRegulation < Sequel::Model
...
-                                                    :complete_abrogation_regulation_role]
+  one_to_one :complete_abrogation_regulation,
+             key: [ :complete_abrogation_regulation_id,
+                    :complete_abrogation_regulation_role ]
+
+  one_to_one :explicit_abrogation_regulation,
+             key: [ :explicit_abrogation_regulation_id,
+                    :explicit_abrogation_regulation_role ]
+
+  one_to_many :generating_measures,
```
[the remark this answers is not in the document]
Comment (mechanicles) [reply app/models/base_regulation.rb]: @matt Not sure about where to move this relation but based on your description over here, currently I am assuming that we need to rename this relationship from `generating_measures` to `measures`.

Not sure how to confirm time machine will respect for `ds.with_actual(MeasureType)` in `Measure` mode.

File: app/models/measure_type.rb [hunk 6]
```diff
@@ -17,7 +17,7 @@ class MeasureType < Sequel::Model
   one_to_one :measure_type_description, key: :measure_type_id,
                                         foreign_key: :measure_type_id
 
-  one_to_many :measures, key: :measure_type,
+  one_to_many :measures, key: :measure_type_id,
```
[the remark this answers is not in the document]
Comment (mechanicles) [reply app/models/measure_type.rb]: Alexey is going to handle this stuff.

File: app/validators/base_regulation_validator.rb (hunk 1 above)
Comment (mechanicles) [reply app/validators/base_regulation_validator.rb]: I will take look into this for handling it based on the time machine.

File: app/validators/base_regulation_validator.rb (hunk 2 above)
Comment (mechanicles) [reply app/validators/base_regulation_validator.rb]: I will check.

**Pull request closed without merging**